source/extensions/http/header_validators/envoy_default/http2_header_validator.cc

```diff
@@ -0,0 +1,415 @@
+      });
+
+  if (!reject_details.empty()) {
+    return {ResponseHeaderMapValidationResult::Action::Reject, reject_details};
+  }
+
+  for (auto& name : drop_headers) {
+    header_map.remove(LowerCaseString(name));
+  }
+
+  return ResponseHeaderMapValidationResult::success();
+}
+
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+Http2HeaderValidator::validateTEHeader(const ::Envoy::Http::HeaderString& value) {
+  // Only allow a TE value of "trailers" for HTTP/2, based on
+  // RFC 9113, https://www.rfc-editor.org/rfc/rfc9113#section-8.2.2:
+  //
+  // The only exception to this is the TE header field, which MAY be present in an HTTP/2 request;
+  // when it is, it MUST NOT contain any value other than "trailers".
+  if (!absl::EqualsIgnoreCase(value.getStringView(), header_values_.TEValues.Trailers)) {
+    return {HeaderEntryValidationResult::Action::Reject, Http2ResponseCodeDetail::get().InvalidTE};
+  }
+
+  return HeaderEntryValidationResult::success();
+}
+
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+Http2HeaderValidator::validateAuthorityHeader(const ::Envoy::Http::HeaderString& value) {
+  // From RFC 3986, https://datatracker.ietf.org/doc/html/rfc3986#section-3.2:
+  //
+  // authority = [ userinfo "@" ] host [ ":" port ]
+  //
+  // HTTP/2 deprecates the userinfo portion of the :authority header. Validate
+  // the :authority header and reject the value if the userinfo is present. This
+  // is based on RFC 9113, https://www.rfc-editor.org/rfc/rfc9113#section-8.3.1:
+  //
+  // ":authority" MUST NOT include the deprecated userinfo subcomponent for "http" or "https"
+  // schemed URIs.
+  //
+  // The host portion can be any valid URI host, which this function does not
+  // validate. The port, if present, is validated as a valid uint16_t port.
+  return validateHostHeader(value);
+}
+
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+Http2HeaderValidator::validateGenericHeaderName(const HeaderString& name) {
+  // Verify that the header name is valid. This also honors the underscore in
+  // header configuration setting.
+  //
+  // From RFC 9110, https://www.rfc-editor.org/rfc/rfc9110.html#section-5.1:
+  //
+  // header-field   = field-name ":" OWS field-value OWS
+  // field-name     = token
+  // token          = 1*tchar
+  //
+  // tchar          = "!" / "#" / "$" / "%" / "&" / "'" / "*"
+  //                / "+" / "-" / "." / "^" / "_" / "`" / "|" / "~"
+  //                / DIGIT / ALPHA
+  //                ; any VCHAR, except delimiters
+  //
+  // For HTTP/2, connection-specific headers must be treated as malformed. From RFC 9113,
+  // https://www.rfc-editor.org/rfc/rfc9113#section-8.2.2:
+  //
+  // Any message containing connection-specific header fields MUST be treated as malformed (Section
+  // 8.1.1).
+  static const absl::node_hash_set<absl::string_view> kRejectHeaderNames = {
+      "transfer-encoding", "connection", "upgrade", "keep-alive", "proxy-connection"};
```

Comment:
Would that approach require this uhv extension to be disabled?  Would the hypothetical separate extension be a fork of this one?  If not, how would that separate extension benefit from all the other functionality of uhv?

I understand how this approach could work if a separate extension would make validation stricter (so that running both extensions would get the desired result), but it's not clear to me how another extension could work to make validation more lenient.

Reply:
The best approach is to derive from these classes and override whatever behavior is needed to be changed and then package it into a separate extension.